File: agent/curiosity.py
"""
agent/curiosity.py — Движок любопытства для YANDI V5.

Система самостоятельно:
- находит пробелы в знаниях
- анализирует убеждения с низкой уверенностью
- формирует вопросы
- приоритизирует неизвестное
- инициирует исследование без пользователя

Теперь с интеграцией с belief_manager.
"""

from __future__ import annotations

import sys
from pathlib import Path
BASE = Path(__file__).parent.parent
sys.path.insert(0, str(BASE))

import json
import time
import uuid
from dataclasses import dataclass, field
from typing import Optional, Dict, Any, List

from agent.belief_manager import get_belief_manager


@dataclass
class Unknown:
    """Неизвестное, которое хочет исследовать система."""
    id: str
    topic: str
    question: str
    priority: float  # 0..1
    reason: str
    created_at: float
    status: str  # pending | exploring | resolved | abandoned
    related_claims: List[str] = field(default_factory=list)
    confidence_gap: float = 0.0
    related_belief_id: Optional[str] = None


class CuriosityEngine:
    """
    Движок любопытства.
    
    Анализирует убеждения и находит, что ещё нужно узнать.
    """
    
    def __init__(self):
        self.unknowns: List[Unknown] = []
        self.exploration_history: List[Dict] = []
        self.belief_manager = get_belief_manager()
    
    def analyze_beliefs(self) -> List[Unknown]:
        """
        Анализировать убеждения и генерировать вопросы.
        """
        new_unknowns = []
        
        # Получаем все активные убеждения
        beliefs = self.belief_manager.get_all_active()
        
        for belief in beliefs:
            # Если уверенность низкая — нужно исследовать
            if belief.confidence < 0.6:
                new_unknowns.append(self._create_unknown(
                    topic=belief.topic,
                    question=f"Почему уверенность в '{belief.statement[:40]}...' составляет {belief.confidence:.2f}?",
                    priority=1.0 - belief.confidence,
                    reason=f"Низкая уверенность в убеждении",
                    confidence_gap=1.0 - belief.confidence,
                    related_belief_id=belief.id
                ))
            
            # Если уверенность средняя — можно углубить
            elif belief.confidence < 0.8:
                new_unknowns.append(self._create_unknown(
                    topic=belief.topic,
                    question=f"Какие ещё доказательства подтверждают '{belief.statement[:40]}...'?",
                    priority=0.5,
                    reason=f"Средняя уверенность — нужны дополнительные данные",
                    confidence_gap=0.3,
                    related_belief_id=belief.id
                ))
        
        # Добавляем новые неизвестные
        for unknown in new_unknowns:
            self._add_unknown(unknown)
        
        return new_unknowns
    
    def analyze_response(
        self,
        query: str,
        epistemic: Dict[str, Any],
        confidence: float,
        evidence_count: int,
        claims: List[Dict],
        trust: str,
    ) -> List[Unknown]:
        """
        Проанализировать ответ и найти неизвестное.
        """
        new_unknowns = []
        
        # 1. Низкая уверенность → нужно больше знать
        if confidence < 0.6:
            new_unknowns.append(self._create_unknown(
                topic=epistemic.get("domain", "general"),
                question=f"Что ещё неизвестно о {query[:50]}?",
                priority=0.7,
                reason=f"Уверенность {confidence:.2f} — нужны дополнительные данные",
                confidence_gap=1.0 - confidence
            ))
        
        # 2. Мало evidence → нужно больше источников
        if evidence_count < 3:
            new_unknowns.append(self._create_unknown(
                topic=epistemic.get("domain", "general"),
                question=f"Какие ещё источники подтверждают или опровергают это?",
                priority=0.6,
                reason=f"Найдено только {evidence_count} источников",
                confidence_gap=0.3
            ))
        
        # 3. Интерпретативный вопрос → нужны разные рамки
        if epistemic.get("testability") == "interpretive":
            new_unknowns.append(self._create_unknown(
                topic="philosophical",
                question="Какие ещё рамки интерпретации существуют?",
                priority=0.5,
                reason="Вопрос интерпретативный — нужно больше перспектив",
                confidence_gap=0.4
            ))
        
        # 4. Противоречия → нужно разрешить
        if trust in ["UNVERIFIED", "PARTIALLY_SUPPORTED"]:
            new_unknowns.append(self._create_unknown(
                topic=epistemic.get("domain", "general"),
                question=f"Почему нет консенсуса по этому вопросу?",
                priority=0.6,
                reason=f"Trust {trust} — нужно понять причины неопределённости",
                confidence_gap=0.5
            ))
        
        # 5. Специфические темы по запросу
        if "сознание" in query.lower():
            new_unknowns.append(self._create_unknown(
                topic="philosophy_of_mind",
                question="Что такое 'трудная проблема сознания'?",
                priority=0.8,
                reason="Связано с текущим запросом",
                confidence_gap=0.7
            ))
        
        if "жизнь" in query.lower() or "смысл" in query.lower():
            new_unknowns.append(self._create_unknown(
                topic="philosophy",
                question="Какие основные философские позиции по смыслу жизни существуют?",
                priority=0.7,
                reason="Связано с текущим запросом",
                confidence_gap=0.6
            ))
        
        # Добавляем новые неизвестные
        for unknown in new_unknowns:
            self._add_unknown(unknown)
        
        return new_unknowns
    
    def _create_unknown(
        self,
        topic: str,
        question: str,
        priority: float,
        reason: str,
        confidence_gap: float = 0.0,
        related_belief_id: Optional[str] = None,
    ) -> Unknown:
        """Создать объект неизвестного."""
        return Unknown(
            id=f"unk_{uuid.uuid4().hex[:8]}",
            topic=topic,
            question=question,
            priority=min(1.0, max(0.0, priority)),
            reason=reason,
            created_at=time.time(),
            status="pending",
            confidence_gap=confidence_gap,
            related_belief_id=related_belief_id,
        )
    
    def _add_unknown(self, unknown: Unknown):
        """Добавить неизвестное, если его ещё нет."""
        # Проверяем дубли
        for existing in self.unknowns:
            if existing.question == unknown.question:
                # Обновляем приоритет
                existing.priority = max(existing.priority, unknown.priority)
                return
        
        self.unknowns.append(unknown)
        self._sort_by_priority()
    
    def _sort_by_priority(self):
        """Сортировать неизвестные по приоритету."""
        self.unknowns.sort(key=lambda x: x.priority, reverse=True)
    
    def get_next_question(self) -> Optional[Unknown]:
        """Получить следующий вопрос для исследования."""
        for unknown in self.unknowns:
            if unknown.status == "pending":
                unknown.status = "exploring"
                return unknown
        return None
    
    def mark_resolved(self, unknown_id: str):
        """Отметить неизвестное как решённое."""
        for unknown in self.unknowns:
            if unknown.id == unknown_id:
                unknown.status = "resolved"
                break
    
    def get_pending(self) -> List[Unknown]:
        """Получить все ожидающие исследования вопросы."""
        return [u for u in self.unknowns if u.status == "pending"]
    
    def get_exploring(self) -> List[Unknown]:
        """Получить все исследуемые вопросы."""
        return [u for u in self.unknowns if u.status == "exploring"]
    
    def get_by_topic(self, topic: str) -> List[Unknown]:
        """Получить неизвестные по теме."""
        return [u for u in self.unknowns if u.topic == topic]
    
    def get_summary(self) -> Dict[str, Any]:
        """Получить сводку любопытства."""
        total = len(self.unknowns)
        pending = len(self.get_pending())
        exploring = len(self.get_exploring())
        resolved = len([u for u in self.unknowns if u.status == "resolved"])
        
        top_questions = [
            {"question": u.question[:80], "priority": u.priority}
            for u in self.unknowns[:5] if u.status != "resolved"
        ]
        
        return {
            "total_unknowns": total,
            "pending": pending,
            "exploring": exploring,
            "resolved": resolved,
            "top_questions": top_questions,
        }
    
    def to_dict(self) -> Dict[str, Any]:
        """Преобразовать в словарь."""
        return {
            "unknowns": [
                {
                    "id": u.id,
                    "topic": u.topic,
                    "question": u.question,
                    "priority": u.priority,
                    "reason": u.reason,
                    "status": u.status,
                    "confidence_gap": u.confidence_gap,
                    "related_belief_id": u.related_belief_id,
                }
                for u in self.unknowns
            ],
            "summary": self.get_summary(),
        }


# Глобальный экземпляр
_curiosity: Optional[CuriosityEngine] = None

def get_curiosity() -> CuriosityEngine:
    global _curiosity
    if _curiosity is None:
        _curiosity = CuriosityEngine()
    return _curiosity


if __name__ == "__main__":
    # Тестирование
    cu = get_curiosity()
    
    # Добавляем убеждения через belief_manager
    bm = get_belief_manager()
    bm.add_belief(
        topic="consciousness",
        statement="Сознание является эмерджентным свойством нейронных сетей",
        confidence=0.55,
        evidence_ids=["ev_001"]
    )
    bm.add_belief(
        topic="abiogenesis",
        statement="Жизнь возникла из неживой материи около 3,5 млрд лет назад",
        confidence=0.9,
        evidence_ids=["ev_002"]
    )
    
    print("=== АНАЛИЗ УБЕЖДЕНИЙ ===")
    new_unknowns = cu.analyze_beliefs()
    print(f"Создано новых неизвестных: {len(new_unknowns)}")
    
    print("\n=== ВСЕ НЕИЗВЕСТНЫЕ ===")
    for u in cu.unknowns:
        print(f"  [{u.priority:.2f}] {u.question[:70]}... ({u.status})")
        print(f"      Причина: {u.reason}")
    
    print("\n=== СВОДКА ===")
    print(cu.get_summary())

File: rustlib/yandi_core/src/curiosity.rs
//! Движок любопытства — перенос `agent/curiosity.py`: система сама находит пробелы в знаниях (убеждения с низкой или средней уверенностью, слабые ответы, интерпретативные вопросы, отсутствие
//! консенсуса), формирует из них вопросы и приоритизирует неизвестное. Список неизвестных живёт в экземпляре, как в оригинале; мост Python↔Rust передаёт его явно.
use serde_json::{json, Value};

use crate::belief_manager;
use crate::ctx::Ctx;
use crate::R;

fn cut(s: &str, n: usize) -> String {
    s.chars().take(n).collect()
}

/// Неизвестное, которое система хочет исследовать.
#[derive(Debug, Clone, PartialEq)]
pub struct Unknown {
    pub id: String,
    /// В оригинале — любое значение (берётся из `epistemic["domain"]` как есть).
    pub topic: Value,
    pub question: String,
    pub priority: f64,
    pub reason: String,
    pub created_at: f64,
    pub status: String,
    pub related_claims: Vec<String>,
    pub confidence_gap: f64,
    pub related_belief_id: Option<String>,
}

impl Unknown {
    pub fn to_json(&self) -> Value {
        json!({
            "id": self.id, "topic": self.topic, "question": self.question, "priority": self.priority, "reason": self.reason, "created_at": self.created_at, "status": self.status,
            "related_claims": self.related_claims, "confidence_gap": self.confidence_gap, "related_belief_id": self.related_belief_id,
        })
    }

    pub fn from_json(v: &Value) -> Unknown {
        Unknown {
            id: v["id"].as_str().unwrap_or("").to_string(),
            topic: v["topic"].clone(),
            question: v["question"].as_str().unwrap_or("").to_string(),
            priority: v["priority"].as_f64().unwrap_or(0.0),
            reason: v["reason"].as_str().unwrap_or("").to_string(),
            created_at: v["created_at"].as_f64().unwrap_or(0.0),
            status: v["status"].as_str().unwrap_or("").to_string(),
            related_claims: v["related_claims"].as_array().map(|a| a.iter().map(|x| x.as_str().unwrap_or("").to_string()).collect()).unwrap_or_default(),
            confidence_gap: v["confidence_gap"].as_f64().unwrap_or(0.0),
            related_belief_id: v["related_belief_id"].as_str().map(String::from),
        }
    }
}

#[derive(Debug, Clone, Default, PartialEq)]
pub struct CuriosityEngine {
    pub unknowns: Vec<Unknown>,
}

impl CuriosityEngine {
    pub fn to_json(&self) -> Value {
        json!({"unknowns": self.unknowns.iter().map(|u| u.to_json()).collect::<Vec<_>>()})
    }

    pub fn from_json(v: &Value) -> CuriosityEngine {
        CuriosityEngine { unknowns: v["unknowns"].as_array().map(|a| a.iter().map(Unknown::from_json).collect()).unwrap_or_default() }
    }

    fn create_unknown(&self, cx: &Ctx, topic: Value, question: String, priority: f64, reason: String, confidence_gap: f64, related_belief_id: Option<String>) -> Unknown {
        Unknown {
            id: format!("unk_{}", cut(&cx.uuid_hex(), 8)),
            topic,
            question,
            priority: priority.min(1.0).max(0.0),
            reason,
            created_at: cx.now_secs(),
            status: "pending".into(),
            related_claims: vec![],
            confidence_gap,
            related_belief_id,
        }
    }

    /// Добавить неизвестное, если такого вопроса ещё нет (иначе — поднять приоритет существующего).
    fn add_unknown(&mut self, unknown: Unknown) {
        for existing in self.unknowns.iter_mut() {
            if existing.question == unknown.question {
                existing.priority = existing.priority.max(unknown.priority);
                return;
            }
        }
        self.unknowns.push(unknown);
        // стабильная сортировка по убыванию приоритета
        self.unknowns.sort_by(|a, b| b.priority.partial_cmp(&a.priority).unwrap_or(std::cmp::Ordering::Equal));
    }

    /// Анализ убеждений: низкая уверенность → «почему», средняя → «какие ещё доказательства».
    pub fn analyze_beliefs(&mut self, cx: &Ctx) -> R<Vec<Unknown>> {
        let mut created: Vec<Unknown> = vec![];
        for b in belief_manager::get_all_active(cx)? {
            if b.confidence < 0.6 {
                created.push(self.create_unknown(
                    cx,
                    json!(b.topic),
                    format!("Почему уверенность в '{}...' составляет {:.2}?", cut(&b.statement, 40), b.confidence),
                    1.0 - b.confidence,
                    "Низкая уверенность в убеждении".into(),
                    1.0 - b.confidence,
                    Some(b.id.clone()),
                ));
            } else if b.confidence < 0.8 {
                created.push(self.create_unknown(
                    cx,
                    json!(b.topic),
                    format!("Какие ещё доказательства подтверждают '{}...'?", cut(&b.statement, 40)),
                    0.5,
                    "Средняя уверенность — нужны дополнительные данные".into(),
                    0.3,
                    Some(b.id.clone()),
                ));
            }
        }
        for u in &created {
            self.add_unknown(u.clone());
        }
        Ok(created)
    }

    /// Анализ ответа: что осталось неизвестным.
    pub fn analyze_response(&mut self, cx: &Ctx, query: &str, epistemic: &Value, confidence: f64, evidence_count: f64, trust: &str) -> R<Vec<Unknown>> {
        if !epistemic.is_object() {
            return Err("AttributeError".into());
        }
        let domain = || epistemic.get("domain").cloned().unwrap_or_else(|| json!("general"));
        let mut created: Vec<Unknown> = vec![];
        if confidence < 0.6 {
            created.push(self.create_unknown(cx, domain(), format!("Что ещё неизвестно о {}?", cut(query, 50)), 0.7, format!("Уверенность {:.2} — нужны дополнительные данные", confidence), 1.0 - confidence, None));
        }
        if evidence_count < 3.0 {
            let shown = if evidence_count.fract() == 0.0 { format!("{}", evidence_count as i64) } else { format!("{evidence_count}") };
            created.push(self.create_unknown(cx, domain(), "Какие ещё источники подтверждают или опровергают это?".into(), 0.6, format!("Найдено только {shown} источников"), 0.3, None));
        }
        if epistemic.get("testability").and_then(|t| t.as_str()) == Some("interpretive") {
            created.push(self.create_unknown(cx, json!("philosophical"), "Какие ещё рамки интерпретации существуют?".into(), 0.5, "Вопрос интерпретативный — нужно больше перспектив".into(), 0.4, None));
        }
        if trust == "UNVERIFIED" || trust == "PARTIALLY_SUPPORTED" {
            created.push(self.create_unknown(cx, domain(), "Почему нет консенсуса по этому вопросу?".into(), 0.6, format!("Trust {trust} — нужно понять причины неопределённости"), 0.5, None));
        }
        let low = yandi_rs::py_text::py_lower(query);
        if low.contains("сознание") {
            created.push(self.create_unknown(cx, json!("philosophy_of_mind"), "Что такое 'трудная проблема сознания'?".into(), 0.8, "Связано с текущим запросом".into(), 0.7, None));
        }
        if low.contains("жизнь") || low.contains("смысл") {
            created.push(self.create_unknown(cx, json!("philosophy"), "Какие основные философские позиции по смыслу жизни существуют?".into(), 0.7, "Связано с текущим запросом".into(), 0.6, None));
        }
        for u in &created {
            self.add_unknown(u.clone());
        }
        Ok(created)
    }

    /// Следующий вопрос для исследования: первый ожидающий становится «исследуется».
    pub fn get_next_question(&mut self) -> Option<Unknown> {
        for u in self.unknowns.iter_mut() {
            if u.status == "pending" {
                u.status = "exploring".into();
                return Some(u.clone());
            }
        }
        None
    }

    pub fn mark_resolved(&mut self, unknown_id: &str) {
        for u in self.unknowns.iter_mut() {
            if u.id == unknown_id {
                u.status = "resolved".into();
                break;
            }
        }
    }

    pub fn get_pending(&self) -> Vec<&Unknown> {
        self.unknowns.iter().filter(|u| u.status == "pending").collect()
    }

    pub fn get_exploring(&self) -> Vec<&Unknown> {
        self.unknowns.iter().filter(|u| u.status == "exploring").collect()
    }

    pub fn get_by_topic(&self, topic: &Value) -> Vec<&Unknown> {
        self.unknowns.iter().filter(|u| &u.topic == topic).collect()
    }

    pub fn get_summary(&self) -> Value {
        let top: Vec<Value> = self.unknowns.iter().take(5).filter(|u| u.status != "resolved").map(|u| json!({"question": cut(&u.question, 80), "priority": u.priority})).collect();
        json!({
            "total_unknowns": self.unknowns.len(), "pending": self.get_pending().len(), "exploring": self.get_exploring().len(),
            "resolved": self.unknowns.iter().filter(|u| u.status == "resolved").count(), "top_questions": top,
        })
    }

    pub fn to_dict(&self) -> Value {
        json!({
            "unknowns": self.unknowns.iter().map(|u| json!({"id": u.id, "topic": u.topic, "question": u.question, "priority": u.priority, "reason": u.reason, "status": u.status, "confidence_gap": u.confidence_gap, "related_belief_id": u.related_belief_id})).collect::<Vec<_>>(),
            "summary": self.get_summary(),
        })
    }
}
